### PR TITLE
Add JSON Field Typecast Support

### PR DESCRIPTION
**Pull request by ekawinataa**
New Classes:

- TypecastedJsonSerializer
- SerializerConfigConverter

**Purpose**:
`TypecastedJsonSerializer` is a decorator for `MessageToTemplatizedJson` and `MessageToJson` serializers.
It enables typecasting for the resulting JSON from the wrapped serializer.
Addresses the issue where default Protobuf to JSON conversion(using `JSONFormat`) treatsing int64 as a String.

Introduces the capability for users to specify JSON paths to be typecasted to INTEGER, LONG, or DOUBLE.
This issue specifically aimed to provide backward compability for users that want to migrate from EGLC to Firehose

**Example Parameter:**
`SERIALIZER_JSON_TYPECAST=[{"jsonPath": "$.someField.nominal", "type": "INTEGER"}]`

it should cast from : 
`{"someField": {"nominal": "12345"}}`

to 
`{"someField": {"nominal":12345}}`

Manually casts the result into a specific type.

**Reason for the PR**
Current Issue: The JsonFormat library converts int64 to String.
User Need: Some users need int64 to be serialized as a JSON integer/number due to strict schema contracts in their Sink.
Solution: A decorator to typecast the JSON output to meet the user's requirements for the HTTP Sink.

**Impact**
This change should only impact the HTTP Sink, specifically the parts which utilize `MessageToTemplatizedJson` and `MessageToJson` serializer

Reference : 
- https://github.com/protocolbuffers/protobuf/issues/1823
- https://github.com/protocolbuffers/protobuf/issues/16270

### CODE REVIEW

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, DEFAULT_PARAMETERS);
+        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(DEFAULT_JSON_MESSAGE);
+        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);
+        typecastedJsonSerializer = new TypecastedJsonSerializer(
+                messageSerializer, serializerConfig
+        );
+    }
+
+    @Test
+    public void serialize_GivenMessageWithQuoteWrappedNumber_ShouldCastToNumber() {
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:47]: change test method name to camelcase like shouldCastToNumberWhenGivenMessageWithQuoteWrappedNumber

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        String processedJsonString = typecastedJsonSerializer.serialize(buildMessage("key", jsonWithNullMappedValue));
+        DocumentContext jsonPath = JsonPath.parse(processedJsonString);
+        JSONArray fieldWithValue = jsonPath.read("$..key");
+        JSONArray integerJsonArray = jsonPath.read("$..long");
+
+        Assertions.assertEquals("value", fieldWithValue.get(0));
+        Assertions.assertNull(integerJsonArray.get(0));
+    }
+
+
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:75]: can you please add more unit tests covering all the corner cases

**Review by sumitaich1998 — COMMENTED** [review]
build is failing due to checkstyle errors

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java (hunk 2 above)
Comment (ekawinataa) [reply src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:75]: sure will do, previously I was using the same JSON payload for multiple cases, let me separate it into multiple tests for more clarity.

File: src/test/java/com/gotocompany/firehose/converter/SerializerConfigConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Test
+    public void convertShouldThrowJsonParseExceptionWhenUnregisteredTypecastingProvided() {
+        String malformedConfigJson = "[{\"jsonPath\": \"$.root.field\", \"type\": \"BIG_INTEGER\"}]";
+
+        Assertions.assertThrows(IllegalArgumentException.class,
+                () -> serializerConfigConverter.convert(null, malformedConfigJson));
+    }
+}
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/converter/SerializerConfigConverterTest.java:43]: please add these additional unit tests 

`   @Test
    public void convertShouldHandleEmptyJsonConfig() {
        String emptyConfigJson = "[]";

        Map<String, Function<String, Object>> result = serializerConfigConverter.convert(null, emptyConfigJson);

        Assertions.assertTrue(result.isEmpty());
    }

    @Test
    public void convertShouldHandleNullJsonConfig() {
        String nullConfigJson = null;

        Map<String, Function<String, Object>> result = serializerConfigConverter.convert(null, nullConfigJson);

        Assertions.assertTrue(result.isEmpty());
    }

    @Test
    public void convertShouldThrowExceptionForUnsupportedDataType() {
        String unsupportedTypeConfigJson = "[{\"jsonPath\": \"$.root.field\", \"type\": \"UNSUPPORTED_TYPE\"}]";

        Assertions.assertThrows(IllegalArgumentException.class,
                () -> serializerConfigConverter.convert(null, unsupportedTypeConfigJson));
    }

    @Test
    public void convertShouldHandleMultipleValidConfigs() {
        String multipleConfigJson = "[{\"jsonPath\": \"$.root.field1\", \"type\": \"LONG\"}, {\"jsonPath\": \"$.root.field2\", \"type\": \"STRING\"}]";

        Map<String, Function<String, Object>> result = serializerConfigConverter.convert(null, multipleConfigJson);

        Function<String, Object> mapper1 = result.get("$.root.field1");
        Function<String, Object> mapper2 = result.get("$.root.field2");

        Assertions.assertNotNull(mapper1);
        Assertions.assertNotNull(mapper2);
        Assertions.assertTrue(mapper1.apply("4") instanceof Long);
        Assertions.assertTrue(mapper2.apply("test") instanceof String);
    }

    @Test
    public void convertShouldThrowExceptionForMissingFieldsInConfig() {
        String missingFieldsConfigJson = "[{\"jsonPath\": \"$.root.field\"}]";

        Assertions.assertThrows(IllegalArgumentException.class,
                () -> serializerConfigConverter.convert(null, missingFieldsConfigJson));
    }`

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        String payload = "{\"key\": \"value\", \"long\":\"1234568129012312\",\"nested\": {\"int\": \"1234\"}, \"double\": \"12.1\"}";
+        String parameters = "[{\"jsonPath\": \"$.key\", \"type\": \"INTEGER\"}]";
+        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, parameters);
+        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);
+        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(payload);
+
+        Assertions.assertThrows(NumberFormatException.class,
+                () -> typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE)));
+    }
+
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:110]: also please add these additional unit tests below - 

```
    @Test
    public void serializeShouldHandleEmptyJsonMessage() {
        String emptyJsonMessage = "{}";
        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(emptyJsonMessage);

        String result = typecastedJsonSerializer.serialize(buildMessage("key", emptyJsonMessage));

        Assertions.assertEquals(JsonPath.parse(emptyJsonMessage).jsonString(), JsonPath.parse(result).jsonString());
    }

    @Test
    public void serializeShouldHandleEmptyJsonPathConfiguration() {
        String parameters = "[]";
        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, parameters);
        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);

        String result = typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE));

        Assertions.assertEquals(JsonPath.parse(DEFAULT_JSON_MESSAGE).jsonString(), JsonPath.parse(result).jsonString());
    }

    @Test
    public void serializeShouldHandleInvalidJsonMessage() {
        String invalidJsonMessage = "{\"key\": \"value\", \"long\":}";
        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(invalidJsonMessage);

        Assertions.assertThrows(com.jayway.jsonpath.InvalidJsonException.class,
                () -> typecastedJsonSerializer.serialize(buildMessage("key", invalidJsonMessage)));
    }

    @Test
    public void serializeShouldHandleNonMatchingJsonPathConfiguration() {
        String parameters = "[{\"jsonPath\": \"$..nonExistentField\", \"type\": \"INTEGER\"}]";
        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, parameters);
        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);

        String result = typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE));

        Assertions.assertEquals(JsonPath.parse(DEFAULT_JSON_MESSAGE).jsonString(), JsonPath.parse(result).jsonString());
    }

    @Test
    public void serializeShouldHandleNestedJsonPathConfiguration() {
        String parameters = "[{\"jsonPath\": \"$..nested.int\", \"type\": \"INTEGER\"}]";
        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, parameters);
        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);

        String result = typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE));
        DocumentContext jsonPath = JsonPath.parse(result);
        JSONArray nestedIntJsonArray = jsonPath.read("$..nested.int");

        Assertions.assertEquals(nestedIntJsonArray.get(0), 1234);
    }

    @Test
    public void serializeShouldHandleMultipleJsonPathConfigurations() {
        String parameters = "[{\"jsonPath\": \"$..int\", \"type\": \"INTEGER\"}, {\"jsonPath\": \"$..double\", \"type\": \"DOUBLE\"}]";
        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, parameters);
        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);

        String result = typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE));
        DocumentContext jsonPath = JsonPath.parse(result);
        JSONArray intJsonArray = jsonPath.read("$..int");
        JSONArray doubleJsonArray = jsonPath.read("$..double");

        Assertions.assertEquals(intJsonArray.get(0), 1234);
        Assertions.assertEquals(doubleJsonArray.get(0), 12.1);
    }
```

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                messageSerializer, serializerConfig
+        );
+
+        String result = typecastedJsonSerializer.serialize(buildMessage("key", DEFAULT_JSON_MESSAGE));
+
+        Assertions.assertEquals(JsonPath.parse(DEFAULT_JSON_MESSAGE).jsonString(), JsonPath.parse(result).jsonString());
+    }
+
+    @Test
+    public void serializeShouldReturnMessageAsItIsWhenJsonPathConfigurationDoesNotMatch() {
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:89]: remove serialize from the starting of the test name , test name should with "should"

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        Assertions.assertTrue(integerJsonArray.get(0) instanceof Integer);
+        Assertions.assertTrue(longJsonArray.get(0) instanceof Long);
+        Assertions.assertTrue(doubleJsonArray.get(0) instanceof Double);
+        Assertions.assertEquals(integerJsonArray.get(0), 1234);
+        Assertions.assertEquals(longJsonArray.get(0), 1234568129012312L);
+        Assertions.assertEquals(doubleJsonArray.get(0), 12.1);
+    }
+
+    @Test
+    public void serializeShouldIgnoreWhenGivenNullMessageValue() {
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:64]: remove serialize from the starting of the test name , test name should with "should"

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        Map<String, Function<String, Object>> property = serializerConfigConverter.convert(null, DEFAULT_PARAMETERS);
+        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(DEFAULT_JSON_MESSAGE);
+        Mockito.when(serializerConfig.getJsonTypecastMapping()).thenReturn(property);
+        typecastedJsonSerializer = new TypecastedJsonSerializer(
+                messageSerializer, serializerConfig
+        );
+    }
+
+    @Test
+    public void serializeShouldCastToNumberWhenGivenMessageWithQuoteWrappedNumberAndMatchingJsonPathConfiguration() {
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:48]: remove serialize from the starting of the test name , test name should with "should"

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        DocumentContext jsonPath = JsonPath.parse(processedJsonString);
+        JSONArray fieldWithValue = jsonPath.read("$..key");
+        JSONArray integerJsonArray = jsonPath.read("$..long");
+
+        Assertions.assertEquals("value", fieldWithValue.get(0));
+        Assertions.assertNull(integerJsonArray.get(0));
+    }
+
+    @Test
+    public void serializeShouldReturnMessageAsItIsWhenNoJsonPathConfigurationGiven() {
```
Comment (sumitaich1998) [inline src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:77]: remove serialize from the starting of the test name , test name should with "should"

File: src/main/java/com/gotocompany/firehose/config/SerializerConfig.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.gotocompany.firehose.config;
+
+import com.gotocompany.firehose.config.converter.SerializerConfigConverter;
+import org.aeonbits.owner.Config;
+
+import java.util.Map;
+import java.util.function.Function;
+
+public interface SerializerConfig extends Config {
```
Comment (sumitaich1998) [inline src/main/java/com/gotocompany/firehose/config/SerializerConfig.java:9]: "There is no need for this separate interface. Can you move this configuration to HttpSinkConfig?"

File: src/main/java/com/gotocompany/firehose/config/SerializerConfig.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    @Config.Key("SERIALIZER_JSON_TYPECAST")
```
Comment (sumitaich1998) [inline src/main/java/com/gotocompany/firehose/config/SerializerConfig.java:10]: after moving you can rename this config and prefix "SINK_HTTP_" to it

**Review by sumitaich1998 — COMMENTED** [review]
add add the new config to the firehose documentation here https://github.com/goto/firehose/blob/main/docs/docs/sinks/http-sink.md and give an example

**Review by sumitaich1998 — COMMENTED** [review]
Also bump up Firehose version, just a minor version bump

File: src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        Assertions.assertEquals(integerJsonArray.get(0), 1234);
+        Assertions.assertEquals(longJsonArray.get(0), 1234568129012312L);
+        Assertions.assertEquals(doubleJsonArray.get(0), 12.1);
+    }
+
+    @Test
+    public void serializeShouldIgnoreWhenGivenNullMessageValue() {
+        String jsonWithNullMappedValue = "{\"key\": \"value\", \"long\":null}";
+        Mockito.when(messageSerializer.serialize(Mockito.any())).thenReturn(jsonWithNullMappedValue);
+        String processedJsonString = typecastedJsonSerializer.serialize(buildMessage("key", jsonWithNullMappedValue));
```
[the remark this answers is not in the document]
Comment (ekawinataa) [reply src/test/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializerTest.java:67]: this one `serializeShouldHandleNestedJsonPathConfiguration` already covered by `shouldCastToNumberWhenGivenMessageWithQuoteWrappedNumberAndMatchingJsonPathConfiguration` but anyway, I'll add the test also ya

File: src/main/java/com/gotocompany/firehose/config/HttpSinkConfig.java [hunk 12]
```diff
@@ -94,4 +96,9 @@ public interface HttpSinkConfig extends AppConfig {
     @DefaultValue("true")
     Boolean getSinkHttpSimpleDateFormatEnable();
 
+    @Key("SINK_HTTP_SERIALIZER_JSON_TYPECAST")
+    @ConverterClass(HttpSinkSerializerJsonTypecastConfigConverter.class)
+    @DefaultValue("[]")
+    Map<String, Function<String, Object>> getJsonTypecastMapping();
```
Comment (sumitaich1998) [inline src/main/java/com/gotocompany/firehose/config/HttpSinkConfig.java:102]: Rename method to getSinkHttpSerializerJsonTypecast()

File: src/main/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializer.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+     * @throws DeserializerException if an error occurs during serialization or typecasting
+     */
+    @Override
+    public String serialize(Message message) throws DeserializerException {
+        String jsonString = messageSerializer.serialize(message);
+        DocumentContext documentContext = JsonPath
+                .using(jsonPathConfiguration)
+                .parse(jsonString);
+
+        for (Map.Entry<String, Function<String, Object>> entry : httpSinkConfig.getJsonTypecastMapping()
```
Comment (sumitaich1998) [inline src/main/java/com/gotocompany/firehose/serializer/TypecastedJsonSerializer.java:56]: refactor this code to the following
```
        serializerConfig.getJsonTypecastMapping().forEach((jsonPath, typecastFunction) -> {
            documentContext.map(jsonPath, (currentValue, configuration) -> 
                Optional.ofNullable(currentValue)
                        .map(value -> typecastFunction.apply(value.toString()))
                        .orElse(null)
```

**Review by sumitaich1998 — CHANGES_REQUESTED** [review]

**Review by sumitaich1998 — APPROVED** [review]
